File: binding/include/multimedia/libqvideoframe.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qvideoframe.html
pub const qvideoframe = struct {

    /// New constructs a new QVideoFrame object.
    pub fn New() QtC.QVideoFrame {
        return qtc.QVideoFrame_new();
    }


    /// New2 constructs a new QVideoFrame object.
    pub fn New2(format: ?*anyopaque) QtC.QVideoFrame {
        return qtc.QVideoFrame_new2(@ptrCast(format));
    }


    /// New3 constructs a new QVideoFrame object.
    pub fn New3(image: ?*anyopaque) QtC.QVideoFrame {
        return qtc.QVideoFrame_new3(@ptrCast(image));
    }


    /// New4 constructs a new QVideoFrame object.
    pub fn New4(other: ?*anyopaque) QtC.QVideoFrame {
        return qtc.QVideoFrame_new4(@ptrCast(other));
    }


    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QVideoFrame_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QVideoFrame_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QVideoFrame_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QVideoFrame_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QVideoFrame_IsValid(@ptrCast(self));
    }

    pub fn PixelFormat(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_PixelFormat(@ptrCast(self));
    }

    pub fn SurfaceFormat(self: ?*anyopaque, ) QtC.QVideoFrameFormat {
        return qtc.QVideoFrame_SurfaceFormat(@ptrCast(self));
    }

    pub fn HandleType(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_HandleType(@ptrCast(self));
    }

    pub fn Size(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QVideoFrame_Size(@ptrCast(self));
    }

    pub fn Width(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_Width(@ptrCast(self));
    }

    pub fn Height(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_Height(@ptrCast(self));
    }

    pub fn IsMapped(self: ?*anyopaque, ) bool {
        return qtc.QVideoFrame_IsMapped(@ptrCast(self));
    }

    pub fn IsReadable(self: ?*anyopaque, ) bool {
        return qtc.QVideoFrame_IsReadable(@ptrCast(self));
    }

    pub fn IsWritable(self: ?*anyopaque, ) bool {
        return qtc.QVideoFrame_IsWritable(@ptrCast(self));
    }

    pub fn MapMode(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_MapMode(@ptrCast(self));
    }

    pub fn Map(self: ?*anyopaque, mode: i32) bool {
        return qtc.QVideoFrame_Map(@ptrCast(self), @intCast(mode));
    }

    pub fn Unmap(self: ?*anyopaque, ) void {
        qtc.QVideoFrame_Unmap(@ptrCast(self));
    }

    pub fn BytesPerLine(self: ?*anyopaque, plane: i32) i32 {
        return qtc.QVideoFrame_BytesPerLine(@ptrCast(self), plane);
    }

    pub fn Bits(self: ?*anyopaque, plane: i32) u8 {
        return @ptrCast(qtc.QVideoFrame_Bits(@ptrCast(self), plane));
    }

    pub fn Bits2(self: ?*anyopaque, plane: i32) u8 {
        return @ptrCast(qtc.QVideoFrame_Bits2(@ptrCast(self), plane));
    }

    pub fn MappedBytes(self: ?*anyopaque, plane: i32) i32 {
        return qtc.QVideoFrame_MappedBytes(@ptrCast(self), plane);
    }

    pub fn PlaneCount(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_PlaneCount(@ptrCast(self));
    }

    pub fn StartTime(self: ?*anyopaque, ) i64 {
        return qtc.QVideoFrame_StartTime(@ptrCast(self));
    }

    pub fn SetStartTime(self: ?*anyopaque, time: i64) void {
        qtc.QVideoFrame_SetStartTime(@ptrCast(self), time);
    }

    pub fn EndTime(self: ?*anyopaque, ) i64 {
        return qtc.QVideoFrame_EndTime(@ptrCast(self));
    }

    pub fn SetEndTime(self: ?*anyopaque, time: i64) void {
        qtc.QVideoFrame_SetEndTime(@ptrCast(self), time);
    }

    pub fn SetRotationAngle(self: ?*anyopaque, angle: i32) void {
        qtc.QVideoFrame_SetRotationAngle(@ptrCast(self), @intCast(angle));
    }

    pub fn RotationAngle(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_RotationAngle(@ptrCast(self));
    }

    pub fn SetRotation(self: ?*anyopaque, angle: i32) void {
        qtc.QVideoFrame_SetRotation(@ptrCast(self), @intCast(angle));
    }

    pub fn Rotation(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame_Rotation(@ptrCast(self));
    }

    pub fn SetMirrored(self: ?*anyopaque, mirrored: bool) void {
        qtc.QVideoFrame_SetMirrored(@ptrCast(self), mirrored);
    }

    pub fn Mirrored(self: ?*anyopaque, ) bool {
        return qtc.QVideoFrame_Mirrored(@ptrCast(self));
    }

    pub fn SetStreamFrameRate(self: ?*anyopaque, rate: f64) void {
        qtc.QVideoFrame_SetStreamFrameRate(@ptrCast(self), rate);
    }

    pub fn StreamFrameRate(self: ?*anyopaque, ) f64 {
        return qtc.QVideoFrame_StreamFrameRate(@ptrCast(self));
    }

    pub fn ToImage(self: ?*anyopaque, ) QtC.QImage {
        return qtc.QVideoFrame_ToImage(@ptrCast(self));
    }

    pub fn SubtitleText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QVideoFrame_SubtitleText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qvideoframe.SubtitleText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetSubtitleText(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QVideoFrame_SetSubtitleText(@ptrCast(self), text_str);
    }

    pub fn Paint(self: ?*anyopaque, painter: ?*anyopaque, rect: ?*anyopaque, options: ?*anyopaque) void {
        qtc.QVideoFrame_Paint(@ptrCast(self), @ptrCast(painter), @ptrCast(rect), @ptrCast(options));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QVideoFrame_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qvideoframe-paintoptions.html
pub const qvideoframe__paintoptions = struct {

    /// New constructs a new QVideoFrame::PaintOptions object.
    pub fn New(other: ?*anyopaque) QtC.QVideoFrame__PaintOptions {
        return qtc.QVideoFrame__PaintOptions_new(@ptrCast(other));
    }


    /// New2 constructs a new QVideoFrame::PaintOptions object.
    pub fn New2(other: ?*anyopaque) QtC.QVideoFrame__PaintOptions {
        return qtc.QVideoFrame__PaintOptions_new2(@ptrCast(other));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QVideoFrame__PaintOptions_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QVideoFrame__PaintOptions_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn BackgroundColor(self: ?*anyopaque, ) QtC.QColor {
        return qtc.QVideoFrame__PaintOptions_BackgroundColor(@ptrCast(self));
    }

    pub fn SetBackgroundColor(self: ?*anyopaque, backgroundColor: QtC.QColor) void {
        qtc.QVideoFrame__PaintOptions_SetBackgroundColor(@ptrCast(self), @ptrCast(backgroundColor));
    }

    pub fn AspectRatioMode(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame__PaintOptions_AspectRatioMode(@ptrCast(self));
    }

    pub fn SetAspectRatioMode(self: ?*anyopaque, aspectRatioMode: i32) void {
        qtc.QVideoFrame__PaintOptions_SetAspectRatioMode(@ptrCast(self), @intCast(aspectRatioMode));
    }

    pub fn PaintFlags(self: ?*anyopaque, ) i32 {
        return qtc.QVideoFrame__PaintOptions_PaintFlags(@ptrCast(self));
    }

    pub fn SetPaintFlags(self: ?*anyopaque, paintFlags: i32) void {
        qtc.QVideoFrame__PaintOptions_SetPaintFlags(@ptrCast(self), @intCast(paintFlags));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QVideoFrame__PaintOptions_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/videoframe.html#types
pub const enums = struct {
    pub const HandleType = enum {
        pub const NoHandle: i32 = 0;
        pub const RhiTextureHandle: i32 = 1;
    };

    pub const MapMode = enum {
        pub const NotMapped: i32 = 0;
        pub const ReadOnly: i32 = 1;
        pub const WriteOnly: i32 = 2;
        pub const ReadWrite: i32 = 3;
    };

    pub const RotationAngle = enum {
        pub const Rotation0: i32 = 0;
        pub const Rotation90: i32 = 90;
        pub const Rotation180: i32 = 180;
        pub const Rotation270: i32 = 270;
    };

    pub const PaintFlag = enum {
        pub const DontDrawSubtitles: i32 = 1;
    };

};

File: binding/src/multimedia/libqvideoframe.cpp
#include <QAbstractVideoBuffer>
#include <QColor>
#include <QImage>
#include <QPainter>
#include <QRectF>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVideoFrame>
#define WORKAROUND_INNER_CLASS_DEFINITION_QVideoFrame__PaintOptions
#include <QVideoFrameFormat>
#include <qvideoframe.h>
#include "libqvideoframe.h"
#include "libqvideoframe.hxx"

QVideoFrame* QVideoFrame_new() {
	 return new QVideoFrame();
}

QVideoFrame* QVideoFrame_new2(const QVideoFrameFormat* format) {
	 return new QVideoFrame(*format);
}

QVideoFrame* QVideoFrame_new3(const QImage* image) {
	 return new QVideoFrame(*image);
}

QVideoFrame* QVideoFrame_new4(std::unique_ptr<QAbstractVideoBuffer> videoBuffer) {
	 return new QVideoFrame(videoBuffer);
}

QVideoFrame* QVideoFrame_new5(const QVideoFrame* other) {
	 return new QVideoFrame(*other);
}

QVideoFrame* QVideoFrame_new6(QAbstractVideoBuffer* buffer, const QVideoFrameFormat* format) {
	 return new QVideoFrame(buffer, *format);
}

void QVideoFrame_Swap(QVideoFrame* self, QVideoFrame* other) {
	self->swap(*other);
}

void QVideoFrame_OperatorAssign(QVideoFrame* self, const QVideoFrame* other) {
	self->operator=(*other);
}

bool QVideoFrame_OperatorEqual(const QVideoFrame* self, const QVideoFrame* other) {
	return self->operator==(*other);
}

bool QVideoFrame_OperatorNotEqual(const QVideoFrame* self, const QVideoFrame* other) {
	return self->operator!=(*other);
}

bool QVideoFrame_IsValid(const QVideoFrame* self) {
	return self->isValid();
}

int QVideoFrame_PixelFormat(const QVideoFrame* self) {
	return self->pixelFormat();
}

QVideoFrameFormat* QVideoFrame_SurfaceFormat(const QVideoFrame* self) {
	return new QVideoFrameFormat(self->surfaceFormat());
}

int QVideoFrame_HandleType(const QVideoFrame* self) {
	return self->handleType();
}

QSize* QVideoFrame_Size(const QVideoFrame* self) {
	return new QSize(self->size());
}

int QVideoFrame_Width(const QVideoFrame* self) {
	return self->width();
}

int QVideoFrame_Height(const QVideoFrame* self) {
	return self->height();
}

bool QVideoFrame_IsMapped(const QVideoFrame* self) {
	return self->isMapped();
}

bool QVideoFrame_IsReadable(const QVideoFrame* self) {
	return self->isReadable();
}

bool QVideoFrame_IsWritable(const QVideoFrame* self) {
	return self->isWritable();
}

int QVideoFrame_MapMode(const QVideoFrame* self) {
	return self->mapMode();
}

bool QVideoFrame_Map(QVideoFrame* self, int mode) {
	return self->map(static_cast<QVideoFrame::MapMode>(mode));
}

void QVideoFrame_Unmap(QVideoFrame* self) {
	self->unmap();
}

int QVideoFrame_BytesPerLine(const QVideoFrame* self, int plane) {
	return self->bytesPerLine(plane);
}

unsigned char* QVideoFrame_Bits(QVideoFrame* self, int plane) {
	return self->bits(plane);
}

const unsigned char* QVideoFrame_Bits2(const QVideoFrame* self, int plane) {
	return self->bits(plane);
}

int QVideoFrame_MappedBytes(const QVideoFrame* self, int plane) {
	return self->mappedBytes(plane);
}

int QVideoFrame_PlaneCount(const QVideoFrame* self) {
	return self->planeCount();
}

long long QVideoFrame_StartTime(const QVideoFrame* self) {
	return self->startTime();
}

void QVideoFrame_SetStartTime(QVideoFrame* self, long long time) {
	self->setStartTime(time);
}

long long QVideoFrame_EndTime(const QVideoFrame* self) {
	return self->endTime();
}

void QVideoFrame_SetEndTime(QVideoFrame* self, long long time) {
	self->setEndTime(time);
}

void QVideoFrame_SetRotationAngle(QVideoFrame* self, int angle) {
	self->setRotationAngle(static_cast<QVideoFrame::RotationAngle>(angle));
}

int QVideoFrame_RotationAngle(const QVideoFrame* self) {
	return self->rotationAngle();
}

void QVideoFrame_SetRotation(QVideoFrame* self, int angle) {
	self->setRotation(static_cast<QtVideo::Rotation>(angle));
}

int QVideoFrame_Rotation(const QVideoFrame* self) {
	return self->rotation();
}

void QVideoFrame_SetMirrored(QVideoFrame* self, bool mirrored) {
	self->setMirrored(mirrored);
}

bool QVideoFrame_Mirrored(const QVideoFrame* self) {
	return self->mirrored();
}

void QVideoFrame_SetStreamFrameRate(QVideoFrame* self, double rate) {
	self->setStreamFrameRate(rate);
}

double QVideoFrame_StreamFrameRate(const QVideoFrame* self) {
	return self->streamFrameRate();
}

QImage* QVideoFrame_ToImage(const QVideoFrame* self) {
	return new QImage(self->toImage());
}

libqt_string QVideoFrame_SubtitleText(const QVideoFrame* self) {
	QString _ret = self->subtitleText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QVideoFrame_SetSubtitleText(QVideoFrame* self, const libqt_string text) {
	self->setSubtitleText(QString::fromUtf8(text.data, text.len));
}

void QVideoFrame_Paint(QVideoFrame* self, QPainter* painter, const QRectF* rect, const QVideoFrame__PaintOptions* options) {
	self->paint(painter, *rect, *options);
}

QAbstractVideoBuffer* QVideoFrame_VideoBuffer(const QVideoFrame* self) {
	return self->videoBuffer();
}

void QVideoFrame_Delete(QVideoFrame* self) {
    delete self;
}

QVideoFrame__PaintOptions* QVideoFrame__PaintOptions_new(const QVideoFrame__PaintOptions* other) {
	 return new QVideoFrame::PaintOptions(*other);
}

QVideoFrame__PaintOptions* QVideoFrame__PaintOptions_new2(QVideoFrame__PaintOptions* other) {
	 return new QVideoFrame::PaintOptions(*other);
}

void QVideoFrame__PaintOptions_CopyAssign(QVideoFrame__PaintOptions* self, QVideoFrame__PaintOptions* other) {
    *self = *other;
}

void QVideoFrame__PaintOptions_MoveAssign(QVideoFrame__PaintOptions* self, QVideoFrame__PaintOptions* other) {
    *self = std::move(*other);
}

QColor* QVideoFrame__PaintOptions_BackgroundColor(const QVideoFrame__PaintOptions* self) {
	return new QColor(self->backgroundColor);
}

void QVideoFrame__PaintOptions_SetBackgroundColor(QVideoFrame__PaintOptions* self, QColor* backgroundColor) {
	self->backgroundColor;
}

int QVideoFrame__PaintOptions_AspectRatioMode(const QVideoFrame__PaintOptions* self) {
	return self->aspectRatioMode;
}

void QVideoFrame__PaintOptions_SetAspectRatioMode(QVideoFrame__PaintOptions* self, int aspectRatioMode) {
	self->aspectRatioMode;
}

int QVideoFrame__PaintOptions_PaintFlags(const QVideoFrame__PaintOptions* self) {
	return self->paintFlags;
}

void QVideoFrame__PaintOptions_SetPaintFlags(QVideoFrame__PaintOptions* self, int paintFlags) {
	self->paintFlags;
}

void QVideoFrame__PaintOptions_Delete(QVideoFrame__PaintOptions* self) {
    delete self;
}

